File: backend/app/schemas/self_assessment_review.py
from uuid import UUID
from typing import Optional, List, Any
from datetime import datetime
from pydantic import BaseModel, Field

from .user import UserDetailResponse


class BucketDecision(BaseModel):
    """Individual bucket decision data for supervisor review"""
    bucket: str  # "performance" or "competency"
    employee_weight: float = Field(..., alias="employeeWeight")
    employee_contribution: float = Field(..., alias="employeeContribution")
    employee_rating: str = Field(..., alias="employeeRating")
    status: str  # "pending", "approved", "rejected"
    supervisor_rating: Optional[str] = Field(None, alias="supervisorRating")
    comment: Optional[str] = None

    model_config = {"populate_by_name": True, "from_attributes": True}


class SelfAssessmentReview(BaseModel):
    """
    Pending self-assessment review item for supervisor approval.
    Represents a bucket-based supervisor feedback record.
    """
    id: UUID  # supervisor_feedback.id
    user_id: UUID = Field(..., alias="userId")  # Employee being reviewed
    period_id: UUID = Field(..., alias="periodId")
    supervisor_id: UUID = Field(..., alias="supervisorId")
    status: str
    bucket_decisions: List[BucketDecision] = Field(default_factory=list, alias="bucketDecisions")
    subordinate: Optional[UserDetailResponse] = None  # Employee details (full user info)
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = {"populate_by_name": True, "from_attributes": True}


class SelfAssessmentReviewList(BaseModel):
    """Paginated list of pending self-assessment reviews"""
    items: List[SelfAssessmentReview]
    total: int
    page: int
    limit: int
    pages: int

    model_config = {"populate_by_name": True, "from_attributes": True}

File: frontend/src/api/types/self-assessment-review.ts
import type { UUID } from './common';

/**
 * Self Assessment Review type definitions
 * These types match the backend Pydantic schemas for pending self-assessment reviews (bucket-based supervisor feedback)
 */

/**
 * User information for review display
 */
export interface ReviewUser {
  id: UUID;
  name?: string | null;
  email?: string | null;
  jobTitle?: string | null;
}

/**
 * Individual bucket decision data for supervisor review
 */
export interface BucketDecision {
  bucket: 'performance' | 'competency';
  employeeWeight: number;
  employeeContribution: number;
  employeeRating: string;
  status: 'pending' | 'approved' | 'rejected';
  supervisorRating?: string | null;
  comment?: string | null;
}

/**
 * Pending self-assessment review item for supervisor approval
 * Represents a bucket-based supervisor feedback record
 */
export interface SelfAssessmentReview {
  id: UUID; // supervisor_feedback.id
  userId: UUID; // Employee being reviewed
  periodId: UUID;
  supervisorId: UUID;
  status: string;
  bucketDecisions: BucketDecision[];
  subordinate?: ReviewUser | null; // Employee details
  createdAt: string; // ISO date string
  updatedAt: string; // ISO date string
}

/**
 * Paginated list of pending self-assessment reviews
 */
export interface SelfAssessmentReviewList {
  items: SelfAssessmentReview[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

/**
 * Request payload for updating bucket decisions
 */
export interface UpdateBucketDecisionsRequest {
  bucketDecisions: BucketDecision[];
  status?: 'draft' | 'submitted';
}
